Can't resolve latest version of a dependency (e.g 'rubygems:colorize:+')

Ran into this problem trying to upgrade to version 1.2.0 of jruby-gradle-plugin in another project that depends on it; that project uses a dependency in this format: 'rubygems:compass:+'.  With jruby 1.2.0 I get an exception from gradle trying to resolve dependencies.  I was able to reproduce the exception with this build.gradle:

``` gradle

import com.github.jrubygradle.JRubyExec

buildscript {
   repositories { jcenter() }

  dependencies {
    classpath "com.github.jruby-gradle:jruby-gradle-plugin:1.2.0"
  }
}

apply plugin: "com.github.jruby-gradle.base"

dependencies {
  jrubyExec "rubygems:colorize:+"
}

task printSomePrettyOutputPlease(type: JRubyExec) {
  description "Execute our nice local print-script.rb"
  script "${projectDir}/print-script.rb"
}
```
Attempting to run the printSomePrettyOutputPlease task will give this exception resolving dependencies:
```
Caused by: org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object '' with class 'java.lang.String' to class 'int'
	at com.github.jrubygradle.internal.GemVersion.compare(GemVersion.groovy:182)
	at com.github.jrubygradle.internal.GemVersion.conflict(GemVersion.groovy:202)
	at com.github.jrubygradle.internal.GemVersion$conflict$0.call(Unknown Source)
	at com.github.jrubygradle.internal.GemVersionResolver.resolve(GemVersionResolver.groovy:74)
	at com.github.jrubygradle.internal.GemVersionResolver$_closure1_closure4.doCall(GemVersionResolver.groovy:28)
	at org.gradle.api.internal.ClosureBackedAction.execute(ClosureBackedAction.java:67)
	at org.gradle.api.internal.artifacts.ivyservice.dependencysubstitution.DefaultDependencySubstitutions$DependencyResolveDetailsWrapperAction.execute(DefaultDependencySubstitutions.java:193)
	at org.gradle.api.internal.artifacts.ivyservice.dependencysubstitution.DefaultDependencySubstitutions$DependencyResolveDetailsWrapperAction.execute(DefaultDependencySubstitutions.java:183)
	at org.gradle.internal.Actions$CompositeAction.execute(Actions.java:76)
	at org.gradle.internal.Actions$CompositeAction.execute(Actions.java:76)
	at org.gradle.api.internal.artifacts.ivyservice.dependencysubstitution.DependencySubstitutionResolver.resolve(DependencySubstitutionResolver.java:40)
	... 110 more
```
It looks like there is a straightforward fix in GemVersion.groovy (it is trying to cast to an int variable after testing to see that the string is not an int).  Not sure if this case should be caught upstream though.

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,5 +1,6 @@
 language: groovy
 sudo: false
+dist: trusty
 
 install:
  - ./gradlew clean

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,6 +1,9 @@
 language: groovy
 sudo: false
-dist: trusty
+addons:
+  apt:
+    packages:
+      - oracle-java8-installer
 
 install:
  - ./gradlew clean

--- a/jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/internal/GemVersion.groovy
+++ b/jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/internal/GemVersion.groovy
@@ -169,8 +169,8 @@ class GemVersion {
             return 1
         }
 
-        int aaObject
-        int bbObject
+        def aaObject
+        def bbObject
 
         if (aDigits[index].isInteger() && bDigits[index].isInteger()) {
             // compare them as number

--- a/jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/internal/GemVersion.groovy
+++ b/jruby-gradle-base-plugin/src/main/groovy/com/github/jrubygradle/internal/GemVersion.groovy
@@ -169,27 +169,13 @@ class GemVersion {
             return 1
         }
 
-        def aaObject
-        def bbObject
-
         if (aDigits[index].isInteger() && bDigits[index].isInteger()) {
             // compare them as number
-            aaObject = aDigits[index] as int
-            bbObject = bDigits[index] as int
+            aDigits[index] as int <=> bDigits[index] as int
         }
         else {
             // compare them as string
-            aaObject = aDigits[index]
-            bbObject = bDigits[index]
-        }
-        if (aaObject < bbObject) {
-            -1
-        }
-        else if (aaObject > bbObject) {
-            1
-        }
-        else {
-            0
+            aDigits[index] <=> bDigits[index]
         }
     }
 

--- a/jruby-gradle-base-plugin/src/test/groovy/com/github/jrubygradle/internal/GemVersionSpec.groovy
+++ b/jruby-gradle-base-plugin/src/test/groovy/com/github/jrubygradle/internal/GemVersionSpec.groovy
@@ -178,4 +178,28 @@ class GemVersionSpec extends Specification {
         expect:
         subject.intersect('[1.2.4, 1.2.4]').conflict() == true
     }
+
+    def "passes non-integer version ranges with no conflicts"() {
+        given:
+        GemVersion subject = new GemVersion('[1.2.bar, 1.2.foo]')
+
+        expect:
+        !subject.conflict()
+    }
+
+    def "fails non-integer version ranges with conflicts"() {
+        given:
+        GemVersion subject = new GemVersion('[1.2.foo, 1.2.bar]')
+
+        expect:
+        subject.conflict()
+    }
+
+    def "it does not throw an exception for a '+' version"() {
+        when:
+        new GemVersion('+').conflict()
+
+        then:
+        notThrown(Exception)
+    }
 }

--- a/jruby-gradle-base-plugin/src/test/groovy/com/github/jrubygradle/internal/GemVersionSpec.groovy
+++ b/jruby-gradle-base-plugin/src/test/groovy/com/github/jrubygradle/internal/GemVersionSpec.groovy
@@ -179,7 +179,7 @@ class GemVersionSpec extends Specification {
         subject.intersect('[1.2.4, 1.2.4]').conflict() == true
     }
 
-    def "passes non-integer version ranges with no conflicts"() {
+    def "finds no conflicts in non-integer version ranges"() {
         given:
         GemVersion subject = new GemVersion('[1.2.bar, 1.2.foo]')
 
@@ -187,7 +187,7 @@ class GemVersionSpec extends Specification {
         !subject.conflict()
     }
 
-    def "fails non-integer version ranges with conflicts"() {
+    def "finds conflicts in non-integer version ranges"() {
         given:
         GemVersion subject = new GemVersion('[1.2.foo, 1.2.bar]')
 
@@ -195,7 +195,7 @@ class GemVersionSpec extends Specification {
         subject.conflict()
     }
 
-    def "it does not throw an exception for a '+' version"() {
+    def "does not throw an exception for a '+' version"() {
         when:
         new GemVersion('+').conflict()
 

